contracts/mosaic/MerklePatriciaProof.sol

```diff
@@ -18,127 +14,79 @@ library MerklePatriciaProof {
         RLP.RLPItem[] memory currentNodeList;
 
         bytes32 nodeKey = root;
         uint pathPtr = 0;
 
         bytes memory path = _getNibbleArray2(encodedPath);
-        if(path.length == 0) {return false;}
-
-        for (uint i=0; i<parentNodes.length; i++) {
-            if(pathPtr > path.length) {return false;}
+        if (path.length == 0) {
+            return false;
+        }
+        //for each node in proof
+        for (uint i = 0; i < parentNodes.length; i++) {
+            if (pathPtr > path.length) {
+                return false;
+            }
 
             currentNode = RLP.toBytes(parentNodes[i]);
-            if(nodeKey != keccak256(currentNode)) {return false;}
+            //check address key for each node should be equal to sha3 of node
+            if (nodeKey != keccak256(currentNode)) {
+                return false;
+            }
             currentNodeList = RLP.toList(parentNodes[i]);
 
-            if(currentNodeList.length == 17) {
-                if(pathPtr == path.length) {
-                    if(keccak256(RLP.toBytes(currentNodeList[16])) == value) {
+            //check if node is branch node
+            if (currentNodeList.length == 17) {
+                //check if whole path is traversed and reached to desired node
+                if (pathPtr == path.length) {
+                    //check if value of node is same as expected account value
+                    if (keccak256(RLP.toBytes(currentNodeList[16])) == value) {
                         return true;
                     } else {
-                      return false;
+                        return false;
                     }
                 }
 
                 uint8 nextPathNibble = uint8(path[pathPtr]);
-                if(nextPathNibble > 16) {return false;}
+                if (nextPathNibble > 16) {
+                    return false;
+                }
+                //select  next node key from branch
                 nodeKey = RLP.toBytes32(currentNodeList[nextPathNibble]);
                 pathPtr += 1;
-            } else if(currentNodeList.length == 2) {
-            pathPtr += _nibblesToTraverse(RLP.toData(currentNodeList[0]), path, pathPtr);
+            } else if (currentNodeList.length == 2) {// check if node is extension or leaf node
+                pathPtr += _nibblesToTraverse(RLP.toData(currentNodeList[0]), path, pathPtr);
 
-                if(pathPtr == path.length) {//leaf node
-                    if(keccak256(RLP.toData(currentNodeList[1])) == value) {
+                if (pathPtr == path.length) {//leaf node
+                    //check if value of node is same as expected account value
+                    if (keccak256(RLP.toData(currentNodeList[1])) == value) {
                         return true;
                     } else {
                         return false;
                     }
                 }
                 //extension node ... test if means that it is empty value
```

Comment:
for extension node the value must be empty but whether it is an extension node is encoded by the hex-prefix encoding of the path in the node